File: frontend/src/api/pronunciation.ts
import api from './index';
import axios from 'axios';
import type { LearningLocale, PronunciationAttempt, PronunciationSession } from '@/types';
import { auth, storage } from '@/config/firebase';
import { getDownloadURL, ref, uploadBytes } from 'firebase/storage';

interface SpeechTokenResponse {
  success: boolean;
  token?: string;
  region?: string;
  expires_at?: string;
  error?: string;
}

interface CreateSessionResponse {
  success: boolean;
  sessionId?: string;
  session?: PronunciationSession;
  error?: string;
}

interface SaveAttemptResponse {
  success: boolean;
  attemptId?: string;
  attempt?: PronunciationAttempt;
  error?: string;
}

interface GetAttemptsResponse {
  success: boolean;
  attempts?: PronunciationAttempt[];
  error?: string;
}

export const getSpeechToken = async (): Promise<{ token: string; region: string; expiresAt: string }> => {
  try {
    const response = await api.post<SpeechTokenResponse>('/azure/speech-token');
    if (response.data.success && response.data.token && response.data.region && response.data.expires_at) {
      return {
        token: response.data.token,
        region: response.data.region,
        expiresAt: response.data.expires_at,
      };
    }
    throw new Error(response.data.error || 'Failed to get speech token');
  } catch (error) {
    if (axios.isAxiosError<SpeechTokenResponse>(error)) {
      const backendMessage = error.response?.data?.error;
      if (backendMessage) {
        throw new Error(backendMessage);
      }
    }
    if (error instanceof Error) {
      throw error;
    }
    throw new Error('Failed to get speech token');
  }
};

export const createPronunciationSession = async (
  locale: LearningLocale,
  options?: { promptSetId?: string; objectiveId?: string }
): Promise<{ sessionId: string }> => {
  const response = await api.post<CreateSessionResponse>('/pronunciation/sessions', {
    locale,
    kind: 'practice',
    promptSetId: options?.promptSetId,
    objectiveId: options?.objectiveId,
  });
  if (response.data.success && response.data.sessionId) {
    return { sessionId: response.data.sessionId };
  }
  throw new Error(response.data.error || 'Failed to create session');
};

export const savePronunciationAttempt = async (
  payload: PronunciationAttempt
): Promise<{ attemptId: string }> => {
  const response = await api.post<SaveAttemptResponse>('/pronunciation/attempts', payload);
  if (response.data.success && response.data.attemptId) {
    return { attemptId: response.data.attemptId };
  }
  throw new Error(response.data.error || 'Failed to save attempt');
};

export const getPronunciationAttempts = async (sessionId: string): Promise<PronunciationAttempt[]> => {
  const response = await api.get<GetAttemptsResponse>(`/pronunciation/sessions/${sessionId}/attempts`);
  if (response.data.success && response.data.attempts) {
    return response.data.attempts;
  }
  throw new Error(response.data.error || 'Failed to load attempts');
};

export const uploadPronunciationAudio = async (payload: {
  sessionId: string;
  promptId: string;
  blob: Blob;
}): Promise<string> => {
  const user = auth.currentUser;
  if (!user) {
    throw new Error('User not authenticated');
  }

  const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
  const filePath = `users/${user.uid}/pronunciation/${payload.sessionId}/${payload.promptId}-${timestamp}.webm`;
  const fileRef = ref(storage, filePath);
  const contentType = payload.blob.type || 'audio/webm';

  await uploadBytes(fileRef, payload.blob, { contentType });
  return getDownloadURL(fileRef);
};

File: backend/routes/pronunciation.py
import os
from datetime import datetime, timedelta

import requests
from flask import Blueprint, current_app, jsonify, request

from backend.route_deps import RouteDeps
from backend.services.compliance import (
    create_consent_event,
    get_retention_policy,
    is_school_voice_context,
    resolve_student_compliance_record,
)


def _resolve_school_pronunciation_policy(deps: RouteDeps, uid: str):
    context = deps.get_school_request_context()
    if not is_school_voice_context(context):
        return None

    org_id = context.active_organization_id
    compliance_record = resolve_student_compliance_record(
        deps,
        org_id=org_id,
        student_uid=uid,
    )
    retention_policy = get_retention_policy(compliance_record.get('retention_policy_id'))
    return {
        'org_id': org_id,
        'context': context,
        'compliance_record': compliance_record,
        'retention_policy': retention_policy,
    }


def create_pronunciation_blueprint(deps: RouteDeps) -> Blueprint:
    bp = Blueprint('pronunciation_routes', __name__)

    @bp.route('/api/azure/speech-token', methods=['POST'])
    @deps.login_required
    def api_speech_token():
        """Issue a short-lived Azure Speech token for the browser SDK."""
        uid = deps.get_current_user_uid()
        school_policy = _resolve_school_pronunciation_policy(deps, uid)
        if school_policy and not school_policy['compliance_record'].get('voice_allowed'):
            create_consent_event(
                deps,
                org_id=school_policy['org_id'],
                student_uid=uid,
                event_type='voice.blocked.pronunciation_token',
                actor_type='student',
                actor_id=uid or '',
                payload={'route': '/api/azure/speech-token'},
            )
            return jsonify({
                'success': False,
                'error': 'Voice consent has not been granted for this student.',
            }), 403

        speech_key = (os.environ.get('AZURE_SPEECH_KEY') or os.environ.get('SPEECH_KEY') or '').strip()
        speech_region = (os.environ.get('AZURE_SPEECH_REGION') or os.environ.get('SPEECH_REGION') or '').strip()

        if not speech_key or not speech_region:
            missing = []
            if not speech_key:
                missing.append('AZURE_SPEECH_KEY')
            if not speech_region:
                missing.append('AZURE_SPEECH_REGION')
            missing_text = ', '.join(missing)
            return jsonify({
                'success': False,
                'error': f'Azure Speech credentials not configured ({missing_text})',
            }), 500

        try:
            response = requests.post(
                f'https://{speech_region}.api.cognitive.microsoft.com/sts/v1.0/issueToken',
                headers={'Ocp-Apim-Subscription-Key': speech_key},
                timeout=10,
            )
            if response.status_code != 200:
                current_app.logger.warning(
                    'Azure speech token request failed with status %s: %s',
                    response.status_code,
                    response.text[:300],
                )
                return jsonify({
                    'success': False,
                    'error': 'Failed to issue Azure Speech token',
                    'provider_error': response.text[:300],
                }), response.status_code

            expires_at = (datetime.utcnow() + timedelta(minutes=9)).isoformat() + 'Z'
            return jsonify({
                'success': True,
                'token': response.text,
                'region': speech_region,
                'expires_at': expires_at,
            })
        except requests.RequestException as e:
            current_app.logger.exception('Azure speech token request exception: %s', e)
            return jsonify({
                'success': False,
                'error': 'Azure Speech service request failed. Please try again shortly.',
            }), 502
        except Exception as e:
            current_app.logger.exception('Unexpected error issuing Azure speech token: %s', e)
            return jsonify({'success': False, 'error': 'Failed to issue Azure Speech token'}), 500

    @bp.route('/api/pronunciation/sessions', methods=['POST'])
    @deps.login_required
    def api_create_pronunciation_session():
        """Create a pronunciation practice session."""
        uid = deps.get_current_user_uid()
        data = request.get_json() or {}

        locale = data.get('locale', 'ko-KR')
        kind = data.get('kind', 'practice')
        prompt_set_id = data.get('promptSetId')
        objective_id = data.get('objectiveId')

        if locale not in deps.allowed_learning_locales:
            return jsonify({'success': False, 'error': 'Invalid locale'}), 400

        try:
            school_policy = _resolve_school_pronunciation_policy(deps, uid)
            if school_policy and not school_policy['compliance_record'].get('voice_allowed'):
                create_consent_event(
                    deps,
                    org_id=school_policy['org_id'],
                    student_uid=uid,
                    event_type='voice.blocked.pronunciation_session',
                    actor_type='student',
                    actor_id=uid or '',
                    payload={'locale': locale, 'kind': kind},
                )
                return jsonify({'success': False, 'error': 'Voice consent has not been granted for this student.'}), 403

            session_id = deps.db.create_pronunciation_session(uid, locale, kind, prompt_set_id, objective_id)
            retention_policy = (
                school_policy['retention_policy']
                if school_policy else get_retention_policy('standard_school')
            )
            return jsonify({
                'success': True,
                'sessionId': session_id,
                'session': {
                    'id': session_id,
                    'locale': locale,
                    'rawAudioStorageAllowed': bool(retention_policy.get('raw_audio_storage_allowed', True)),
                    'retentionPolicyId': retention_policy.get('id', 'standard_school'),
                    'voiceAllowed': True,
                },
            })
        except Exception as e:
            return jsonify({'success': False, 'error': str(e)}), 500

    @bp.route('/api/pronunciation/attempts', methods=['POST'])
    @deps.login_required
    def api_save_pronunciation_attempt():
        """Save a pronunciation assessment attempt."""
        uid = deps.get_current_user_uid()
        data = request.get_json() or {}

        session_id = data.get('sessionId')
        prompt_id = data.get('promptId')
        reference_text = data.get('referenceText')
        recognized_text = data.get('recognizedText', '')
        locale = data.get('locale')
        objective_id = data.get('objectiveId')
        scores = data.get('scores', {})
        words = data.get('words', [])
        raw_result = data.get('rawResult')
        audio_url = data.get('audioUrl')

        if not session_id:
            return jsonify({'success': False, 'error': 'sessionId is required'}), 400
        if not prompt_id:
            return jsonify({'success': False, 'error': 'promptId is required'}), 400
        if not reference_text:
            return jsonify({'success': False, 'error': 'referenceText is required'}), 400
        if not locale or locale not in deps.allowed_learning_locales:
            return jsonify({'success': False, 'error': 'Invalid locale'}), 400

        try:
            school_policy = _resolve_school_pronunciation_policy(deps, uid)
            if school_policy and not school_policy['compliance_record'].get('voice_allowed'):
                create_consent_event(
                    deps,
                    org_id=school_policy['org_id'],
                    student_uid=uid,
                    event_type='voice.blocked.pronunciation_attempt',
                    actor_type='student',
                    actor_id=uid or '',
                    payload={'sessionId': session_id, 'promptId': prompt_id},
                )
                return jsonify({'success': False, 'error': 'Voice consent has not been granted for this student.'}), 403

            session = deps.db.get_pronunciation_session(uid, session_id)
            if not session:
                return jsonify({'success': False, 'error': 'Session not found'}), 404

            allowed_audio_url = audio_url
            if school_policy and not school_policy['retention_policy'].get('raw_audio_storage_allowed', True):
                allowed_audio_url = None
                if audio_url:
                    create_consent_event(
                        deps,
                        org_id=school_policy['org_id'],
                        student_uid=uid,
                        event_type='retention.audio_storage_suppressed',
                        actor_type='student',
                        actor_id=uid or '',
                        payload={'sessionId': session_id, 'promptId': prompt_id},
                    )

            attempt_id = deps.db.add_pronunciation_attempt(uid, session_id, {
                'prompt_id': prompt_id,
                'objective_id': objective_id,
                'reference_text': reference_text,
                'recognized_text': recognized_text,
                'locale': locale,
                'scores': scores,
                'words': words,
                'raw_result': raw_result,
                'audio_url': allowed_audio_url,
            })
            return jsonify({'success': True, 'attemptId': attempt_id})
        except Exception as e:
            return jsonify({'success': False, 'error': str(e)}), 500

    @bp.route('/api/pronunciation/sessions/<session_id>/attempts', methods=['GET'])
    @deps.login_required
    def api_get_pronunciation_attempts(session_id):
        """Get pronunciation attempts for a session."""
        uid = deps.get_current_user_uid()
        objective_id = request.args.get('objectiveId')
        try:
            session = deps.db.get_pronunciation_session(uid, session_id)
            if not session:
                return jsonify({'success': False, 'error': 'Session not found'}), 404

            attempts = deps.db.get_pronunciation_attempts(uid, session_id, limit=50, objective_id=objective_id)
            return jsonify({'success': True, 'attempts': attempts})
        except Exception as e:
            return jsonify({'success': False, 'error': str(e)}), 500

    return bp
